### ISSUE TITLE
[feature request] Make tags extendable (with solution)

### ISSUE THREAD

**Issue by munxar** [responsiv/campaign-plugin#44]
I want to extend the available tags in the message editor. 
With three little events it was possible to totally customize the tags. All modifications are done in **Responsiv\Campaign\Models\Message**

```php
    public static function getAvailableTags()
    {
        $tags = [
            'first_name'      => "Subscriber's first name",
            'last_name'       => "Subscriber's last name",
            'email'           => "Subscriber's email address",
            'unsubscribe_url' => "Link to unsubscribe from emails",
            'browser_url'     => "Link to open web-based version"
        ];

        // extend tags that are visible in the message editor 
        Event::fire('responsiv.campaign.extendAvailableTags', [&$tags]);

        return $tags;
    }
```

```php
    public function buildTagData($subscriber)
    {
        $data = [];

        $data['first_name'] = $subscriber->first_name;
        $data['last_name'] = $subscriber->last_name;
        $data['email'] = $subscriber->email;
        $data['unsubscribe_url'] = $this->getBrowserUrl($subscriber).'?unsubscribe=1';
        $data['browser_url'] = $this->getBrowserUrl($subscriber);
        $data['tracking_pixel'] = $this->getTrackingPixelImage($subscriber);
        $data['tracking_url'] = $this->getBrowserUrl($subscriber).'.png';

        // extend tag data send with the email
        Event::fire('responsiv.campaign.extendTagData', [&$data, $subscriber]);

        return $data;
    }
```

```php
    public function renderForPreview()
    {
        $content = $this->content_html ?: $this->renderTemplate();
        $parser = new TextParser;

        $data = [];
        $data['first_name'] = 'Test';
        $data['last_name'] = 'Person';
        $data['email'] = 'test@email.tld';
        $data['unsubscribe_url'] = 'javascript:;';
        $data['browser_url'] = 'javascript:;';
        $data['tracking_pixel'] = '';
        $data['tracking_url'] = 'javascript:;';
        
        // extend preview data
        Event::fire('responsiv.campaign.extendPreviewData', [&$data]);
        
        $result = $parser->parseString($content, $data);

        // Inject base target
        $result = str_replace(
            '</head>',
            '<base target="_blank" />' . PHP_EOL . '</head>',
            $result
        );

        return $result;
    }
```

Now my plugins boot method can do things like that (assume that the subscriber model was extended with a salutation attribute):
```php
    public function boot()
    {
        Event::listen('responsiv.campaign.extendAvailableTags', function(&$tags) {   
            // add a new attribute to the sidebar
            $tags['salutation'] = 'Salutation';
        });

        Event::listen('responsiv.campaign.extendTagData', function(&$data, $subscriber) {
            $data['salutation'] = $subscriber->salutation;
        });

        Event::listen('responsiv.campaign.extendPreviewData', function(&$data) {
            $data['salutation'] = 'Hello';
        });
   }
```

It would be helpful if such an extension would be added to the plugin.

**Comment by damsfx** [comment]
+1
Very useful to add a civility field `Mr, Ms or Mrs ` for more personalized emails.

**Comment by grahambrown11** [comment]
@daftspunk any chance this can be added?
I was going to start applying a similar patch and fortunately came across @munxar post here and he's even given a solution

**Comment by LukeTowers** [comment]
@grahambrown11 @munxar @damsfx how about the following functionality:
```php
    /**
     * Register tags for the Responsiv.Campaign plugin to use
     *
     * @return array
     */
    public function registerCampaignTags()
    {
        return [
            'first_name'      => [
                'description' => "Subscriber's first name",
                'preview'     => "Test",
                'value'       => function ($message, $subscriber) {
                    return $subscriber->first_name;
                },
            ],
            'last_name'       => [
                'description' => "Subscriber's last name",
                'preview'     => "Person",
                'value'       => function ($message, $subscriber) {
                    return $subscriber->last_name;
                },
            ],
            'email'           => [
                'description' => "Subscriber's email address",
                'preview'     => "test@email.tld",
                'value'       => function ($message, $subscriber) {
                    return $subscriber->email;
                },
            ],
            'unsubscribe_url' => [
                'description' => "Link to unsubscribe from emails",
                'preview'     => "javascript:;",
                'value'       => function ($message, $subscriber) {
                    return $message->getBrowserUrl($subscriber).'?unsubscribe=1';
                },
            ],
            'browser_url'     => [
                'description' => "Link to open web-based version",
                'preview'     => "javascript:;",
                'value'       => function ($message, $subscriber) {
                    return $message->getBrowserUrl($subscriber);
                },
            ],
            'tracking_pixel'  => [
                'preview'     => '',
                'value'       => function ($message, $subscriber) {
                    return $message->getTrackingPixelImage($subscriber);
                },
            ],
            'tracking_url'    => [
                'preview'     => 'javascript:;',
                'value'       => function ($message, $subscriber) {
                    return $message->getBrowserUrl($subscriber).'.png';
                },
            ],
        ];
    }
```

The idea being that you'd have this function in your Plugin.php and this plugin would go through all the plugins to register any campaign tags that are provided by external plugins.

**Comment by grahambrown11** [comment]
@LukeTowers I don't mind how it's done as long as we're able to extend the plugin by adding adding our own custom tags.
This approach does seem better as its encapsulated in one method, instead of 3, I'm a little lost as to how our plugin would extend though, is there some undocumented  `register` function magic that's not found at [Extending Plugins](https://octobercms.com/docs/plugin/extending)?
Last thought - I don't want to replace what the plugin is providing as tags, I want to add extra ones so unless the plugin will `array_merge` theirs and ours I'd assume the tags array would be passed in to the extension function

**Comment by LukeTowers** [comment]
@grahambrown11 the idea is that they would be merged. 

As to how the feature would be implemented technically, the register*() method pattern is handled by each case that supports it, so it's not built in magic as such, just magic that's introduced by each use of it. See https://github.com/octobercms/october/blob/master/modules/system/classes/MailManager.php#L340-L356 and https://github.com/rainlab/pages-plugin/blob/739ad6ef1215c269cf407aeb8318ca8939decc80/classes/SnippetManager.php#L196-L216

**Comment by LukeTowers** [comment]
The only difference between the two methods (register vs. event handlers) would be being able to be a bit more selective about when the tags are being applied as the register method would just always be applied without the potential to filter by looking at the message (or even the subscriber, although I don't think you'd want to do that). We could mitigate that by provided the message object to the register method so that the filtering could be done within the register method.

I just don't know if that would be a useful feature or not. Thoughts?

**Comment by damsfx** [comment]
My preference goes to the "register" method...   
One place to rule them all.

I think it's more efficient than having to watch 3 different events

**Comment by grahambrown11** [comment]
> As to how the feature would be implemented technically, the register*() method pattern is handled by each case that supports it, so it's not built in magic as such, just magic that's introduced by each use of it. See https://github.com/octobercms/october/blob/master/modules/system/classes/MailManager.php#L340-L356 and https://github.com/rainlab/pages-plugin/blob/739ad6ef1215c269cf407aeb8318ca8939decc80/classes/SnippetManager.php#L196-L216

Thanks for the explanation, I will keep this pattern in mind for future

> The only difference between the two methods (register vs. event handlers) would be being able to be a bit more selective about when the tags are being applied as the register method would just always be applied without the potential to filter by looking at the message (or even the subscriber, although I don't think you'd want to do that). We could mitigate that by provided the message object to the register method so that the filtering could be done within the register method.
> 
> I just don't know if that would be a useful feature or not. Thoughts?

As for the need to filter tags - I don't think it's required, the once off registering approach should works for me, when it comes to resolving the tag the filtering can be done there with your closure pattern

**Comment by LukeTowers** [comment]
Sounds good. If you need this immediately consider sponsoring me to bump it up in priority or ping me if I haven't gotten to this in a month. Pretty busy with other work right now.

**Comment by grahambrown11** [comment]
> Sounds good. If you need this immediately consider sponsoring me to bump it up in priority or ping me if I haven't gotten to this in a month. Pretty busy with other work right now.

What!!! 
@LukeTowers I respect that you work on OctoberCMS and require a revenue source, but please understand I recently paid $100 for this plugin. Personally I would've expected @daftspunk to be more responsive to customers with that price tag, this issue was filed more than 2 years ago.
If the plugin was free or less than $20 maybe I would find your request less offensive as I did.
@daftspunk if you've got @LukeTowers help to maintain this then you to share some of the revenue with him, or at least sponsoring him for this change!

**Comment by LukeTowers** [comment]
@grahambrown11 this is an issue with how the marketplace licensing system works currently. All products are sold as-is with no guarantee of ongoing support / updates from the author. See https://github.com/octobercms/october/issues/4662 for an ongoing discussion for improvements that can be made in this regard.

I'm perfectly happy to add this for free, as I'll probably end up using this myself at some point, but if I do that it'll be on my schedule when I can get around to it. If you'd like to reach out directly to @daftspunk to ask him to implement the change as a customer request then you can do so here: https://octobercms.com/author/Responsiv. Please note however that he's a CTO of a startup and barely has time to check in with the main project from time to time so he's unlikely to have time to implement this, although it never hurts to ask nicely. 

When it comes to getting his attention on GitHub issues pretty much the only way to do so is by reaching out through other channels as he gets way too many notifications through GitHub to address any of them unless someone has contacted him through another channel.

**Comment by grahambrown11** [comment]
Thanks @LukeTowers, sorry for the harshness in my response - I was a little surprised, I do thank you for your time and responsiveness. I respect you have lots on your schedule, I do too, and apologise if I was imposing my schedule on yours ;-). For now I'll apply @munxar patch to my server and look forward to your implementation when you get time.
I will ping Samuel via the OctoberCMS site as an FYI

**Comment by LukeTowers** [comment]
@grahambrown11 FYI I just found `System\Classes\PluginManager::getRegistrationMethodValues()` which can do what I linked to much easier: https://github.com/octobercms/october/blob/master/modules/system/classes/PluginManager.php#L456-L479

**Comment by damsfx** [comment]
> 
> 
> ... ping me if I haven't gotten to this in a month. Pretty busy with other work right now.

Ping !   ;o)  
I hope the schedule's a little less cluttered.

**Comment by LukeTowers** [comment]
@damsfx it's still on my list but unfortunately still pretty busy :) Keep pinging me, I will get to it eventually

**Comment by daftspunk** [comment]
Removed all the off-topic conversations. This feature has been implemented in v1.3.1, see the updated readme for details.

Thanks!
